packages/code-signing-certificates/src/index.ts

```diff
@@ -0,0 +1,144 @@
+import assert from 'assert';
+import { md, pki as PKI, random, util } from 'node-forge';
+
+// a hexString is considered negative if it's most significant bit is 1
+// because serial numbers use ones' complement notation
+// this RFC in section 4.1.2.2 requires serial numbers to be positive
+// http://www.ietf.org/rfc/rfc5280.txt
+function toPositiveHex(hexString: string) {
+  let mostSiginficativeHexAsInt = parseInt(hexString[0], 16);
+  if (mostSiginficativeHexAsInt < 8) {
+    return hexString;
+  }
+  mostSiginficativeHexAsInt -= 8;
+  return mostSiginficativeHexAsInt.toString() + hexString.substring(1);
+}
+
+export type CertificateAttrs = PKI.CertificateField[];
+
+export function generateKeyPair(): PKI.rsa.KeyPair {
+  return PKI.rsa.generateKeyPair();
+}
+
+export function convertKeyPairToPEM(
+  keyPair: PKI.rsa.KeyPair
+): { privateKeyPEM: string; publicKeyPEM: string } {
+  return {
+    privateKeyPEM: PKI.privateKeyToPem(keyPair.privateKey),
+    publicKeyPEM: PKI.publicKeyToPem(keyPair.publicKey),
+  };
+}
+
+export function convertCertificateToPEM(cert: PKI.Certificate): string {
+  return PKI.certificateToPem(cert);
+}
+
+type GenerateParameters = {
+  /**
+   * Public/private key pair generated via {@link generateKeyPair}.
+   */
+  keyPair: PKI.rsa.KeyPair;
+
+  /**
+   * Certificate validity range start.
+   */
+  validityNotBefore: Date;
+
+  /**
+   * Certificate validity range end.
+   */
+  validityNotAfter: Date;
+
+  /**
+   * CN issuer and subject Distinguished Name (DN).
+   * Used for both issuer and subject in the case of self-signed certificates.
+   */
+  commonName: string;
+};
+
+/**
+ * Generate a self-signed code-signing certificate for use with expo-updates.
+ * Note that while certificate chains may be supported at some point in expo-updates, for now
+ * only self-signed certificates are supported.
+ *
+ * @returns PKI.Certificate valid for expo-updates code signing
+ */
+export function generateSelfSignedCodeSigningCertificate({
+  keyPair: { publicKey, privateKey },
+  validityNotBefore,
+  validityNotAfter,
+  commonName,
+}: GenerateParameters): PKI.Certificate {
+  const cert = PKI.createCertificate();
+  cert.publicKey = publicKey;
+  cert.serialNumber = toPositiveHex(util.bytesToHex(random.getBytesSync(9)));
+
+  assert(
+    validityNotAfter >= validityNotBefore,
+    'validityNotAfter must be later than validityNotBefore'
+  );
+  cert.validity.notBefore = validityNotBefore;
+  cert.validity.notAfter = validityNotAfter;
+
+  const attrs = [
+    {
+      name: 'commonName',
+      value: commonName,
+    },
+  ];
+  cert.setSubject(attrs);
+  cert.setIssuer(attrs);
```

Comment:
Is this the standard values for self signed?

Reply:
It's standard for subject and issuer to be the same. That being said, the only required attribute seems to be commonName, but it is common for other fields to be specified. In this case it doesn't matter, but depending on where we go with expo go certificates and child certificates, we may use these fields more.